pandas/core/groupby/generic.py

```diff
@@ -602,14 +602,18 @@ def value_counts(
         ids, _, _ = self.grouper.group_info
         val = self.obj._values
 
+        names = self.grouper.names + [self.obj.name]
+
         def apply_series_value_counts():
-            return self.apply(
```

Comment:
Could you take this opportunity to refactor and inline this logic like

```
if (bins is not None and np.iterable(bins) or is_categorical_dtype(val.dtype):
    # comments
    s = self.apply(...)
    s.index.names = ...
```

Also will s always have a MultiIndex here?

Reply:
If I run the test code with the following change 

```python
    if is_categorical_dtype(val.dtype) or (bins and np.iterable(bins)):
        # scalar bins cannot be done at top level
        # in a backward compatible way
        # GH38672
        s = self.apply(
            Series.value_counts,
            normalize=normalize,
            sort=sort,
            ascending=ascending,
            bins=bins,
        )
        print(s.index)
        s.index.names = names
```

I get these results

```python
df = pd.DataFrame(
    {
        "gender": ["female"],
        "country": ["US"],
    }
)
df["gender"] = df["gender"].astype("category")
result2 = df.groupby("country")["gender"].value_counts()
MultiIndex([('US', 'female')],
           names=['country', None])
```
And I think the fact that we are inside a group by should always ensure that we are always using MultiIndex.

Running the following code calls a different method. Specifically, `pandas.core.algorithms.value_counts`.